### PR TITLE
HTTP header sanitizing fixes

### PR DESCRIPTION
**Pull request by mattklein123**
Fixes https://github.com/lyft/envoy/issues/1607

1) Clarify header sanitizing in docs
2) A couple of fixes and tests for headers that should be
   sanitized but are not.

### CODE REVIEW

File: docs/configuration/http_conn_man/header_sanitizing.rst [hunk 1]
```diff
@@ -0,0 +1,34 @@
+.. _config_http_conn_man_header_sanitizing:
+
+HTTP header sanitizing
+======================
+
+For security reasons, Envoy will "sanitize" various incoming HTTP headers depending on whether the
+request is an internal or external request. The sanitizing action depends on the header and may
+result in addition, removal, or modification. Ultimately, whether the request is considered internal
+or external is governed by the :ref:`x-forwaded-for <config_http_conn_man_headers_x-forwarded-for>`
```
Comment (htuch) [inline docs/configuration/http_conn_man/header_sanitizing.rst]: Nit: typo forwaded.

File: docs/configuration/http_conn_man/header_sanitizing.rst [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+* :ref:`x-forwarded-proto <config_http_conn_man_headers_x-forwarded-proto>`
+* :ref:`x-request-id <config_http_conn_man_headers_x-request-id>`
+* :ref:`x-envoy-internal <config_http_conn_man_headers_x-envoy-internal>`
+* :ref:`x-envoy-force-trace <config_http_conn_man_headers_x-envoy-force-trace>`
+* :ref:`x-envoy-external-address <config_http_conn_man_headers_x-envoy-external-address>`
+* :ref:`x-envoy-downstream-service-node <config_http_conn_man_headers_downstream-service-node>`
+* :ref:`x-envoy-downstream-service-cluster
+  <config_http_conn_man_headers_downstream-service-cluster>`
+* :ref:`x-envoy-retry-on <config_http_filters_router_x-envoy-retry-on>`
+* :ref:`x-envoy-grpc-retry-on <config_http_filters_router_x-envoy-grpc-retry-on>`
```
Comment (htuch) [inline docs/configuration/http_conn_man/header_sanitizing.rst]: Sort alpha?

File: docs/configuration/http_conn_man/header_sanitizing.rst [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+* :ref:`x-envoy-max-retries <config_http_filters_router_x-envoy-max-retries>`
+* :ref:`x-envoy-upstream-alt-stat-name <config_http_filters_router_x-envoy-upstream-alt-stat-name>`
+* :ref:`x-envoy-upstream-rq-timeout-ms <config_http_filters_router_x-envoy-upstream-rq-timeout-ms>`
+* :ref:`x-envoy-upstream-rq-per-try-timeout-ms
```
Comment (htuch) [inline docs/configuration/http_conn_man/header_sanitizing.rst]: Random thought that might not be practical; would it make sense to have distinct header namespace for headers that are to be filtered on external req? I.e. you could just strip `x-envoy-req-*`.

Comment (mattklein123) [reply docs/configuration/http_conn_man/header_sanitizing.rst]: I would like a better way of grouping things that need to be sanitized, but not sure it's reasonable to change these things now.

**Comment by mattklein123** [comment]
@htuch updated

**Pull request merged**